### ISSUE TITLE
click to simulate an active div tag

### ISSUE THREAD

**Issue by PriyalSayonetech** [airbnb/enzyme#1426]
Hi guys, I'm here with another issue.
Its reactjs. my <div> tags are similar to following:-
`<div class="conversationItem"><span>piloy</span>.....</div>`
Onclick, the div will change to following:-
`<div class="conversationItem active"><span>piloy</span>.....</div>`

This is the test i wrote to simulate the click activity:-
```
            wrapper.find('div.conversationItem').simulate('click');
            wrapper.update();
            expect(wrapper.find('div.conversationItem.active')).to.have.length(1);
```
But its not working as i expected.
`div.conversationItem.active` is never found. it's count is 0.
People, Please help me to overcome this.

**Comment by ljharb** [comment]
I don't see any `onClick` on your div; what's "simulate" triggering?

In general, `simulate` should be avoided; it doesn't faithfully simulate anything.

**Comment by PriyalSayonetech** [comment]
@ljharb  sorry actually there is an onclick function:-
`<div class="conversationItem" onClick={this.handleClick}><span>piloy</span>.....</div>`
and --
handleClick: React.PropTypes.func.isRequired

**Comment by ljharb** [comment]
What happens if you do `wrapper.find('div.conversationItem').prop('onClick')(); wrapper.update()`?

**Comment by PriyalSayonetech** [comment]
@ljharb , thanks but Still no result. i tried with props method too.

i wrote something like this, i don't know whether its right to write like this. 

```
        const handleClickStub = sinon.spy();
        const wrapper=shallow(<ConversationItem conversation={conversation} handleClick={handleClickStub} />);        
        wrapper.find('div.conversationItem').simulate('click');
        expect(handleClickStub).to.have.property('callCount', 1);
```

but with above, my condition is passed. but `expect(wrapper.find('div.conversationItem.active')).to.have.length(1);` is still failing.

**Comment by artiebits** [comment]
Because if you use React as you mentioned you have to use `className` instead of `class` in components https://reactjs.org/docs/dom-elements.html#classname. That's why `conversationItem`  is never found

**Comment by blackpost38** [comment]
@ljharb, I've seen that you said `.simulate` just calls prop about event. Also, you said `.simulate` should be avoided in general. 

I have a question about this one. When is `.simulate` useful? Also, if `.simulate` should be avoided, how about deprecating it?

**Comment by ljharb** [comment]
In my opinion, it is never useful. I would like to deprecate it, and hope to remove it in v4, but that requires discussion with the other enzyme maintainers, and we're not in a rush to do so.

**Comment by matteoantoci** [comment]
I've actually managed to migrate from v2 to v3 with this dirty hack:

```javascript
const DEFAULT_EVENT = { target: {} };

const ACTIONS = {
  click: 'onClick',
  blur: 'onBlur',
  change: 'onChange',
  keyPress: 'onKeyPress',
  focus: 'onFocus',
};

function simulate(action, event = DEFAULT_EVENT) {
  this.prop(ACTIONS[action])(event);
}

Enzyme.ReactWrapper.prototype.simulate = simulate;
```

Just put it near you Enzyme configuration and customize the supported `ACTIONS` object.

**Comment by mir3z** [comment]
This is not really related to this topic but I'm in the middle of migration from v2 to v3 and I can't understand one thing. Why do I have to call `update` after calling `simulate`? I didn't have to do that in v2. but now most of my tests are failing because of this missing `update`.  I can't see any pattern here. Sometimes I need to call it, sometimes not, regardless of renderer I'm using.

**Comment by ljharb** [comment]
@matteoantoci it's not a good idea to monkeypatch things like that; it might work as a very temporary solution, but that doesn't mean it's something that should be recommended widely.

@mir3z if you're seeing inconsistencies, please file them as a separate issue.

**Comment by matteoantoci** [comment]
@ljharb I completely agree with you, though, that dirty monkeypatch let my tests pass after the v3 upgrade. That means there's braking change not documented in the "migration guide".

BTW, I think Enzyme's team should deprecate `simulate` ASAP.

**Comment by ljharb** [comment]
@matteoantoci it'd be great to get a PR that updates the migration guide to mention this difference.
